wireapp/wire-ios-sync-engine
# feat(api-version): use api version in request strategies

----
#### PR Submission Checklist for internal contributors

- The **PR Title**

- The **PR Description**
----

# What's new in this PR?

In this PR we update the request creation logic in order to use the current api version (`APIVersion.current`) instead of using `v0` by default. Since `APIVersion.current` returns an optional `APIVersion`, we need to handle the `nil` case.

Some of the functions used to create requests receive a completion handler. In these cases, if the version is `nil` we pass a `failure` result into the completion handler. Failures (mostly) need an `Error` as associated value. When defined, we associate the `unknown` error. Otherwise, we create a special case for the `missingAPIVersion` error. 

The right strategy to handle the nullability of `APIVersion.current` isn't entirely clear yet. As well as the necessity to handle nullability, since we expect the API version to always be fetched when the app enters the foreground.

Another change concerns the `FederationAware` request strategies. There we cannot simply keep the existing behaviour. So we switch over the API version. For `v0` we keep creating requests for legacy endpoints. For `v1`, since federation endpoints are supported regardless of whether or not federation is enabled, we create requests for federation endpoints. However, when federation is disabled, we use the domain from `APIVersion`. 

Now, in practice, we expect the domain from the subject of the request to be `nil` when federation is disabled. So, we don't directly check the federation flag. Instead, if the domain is `nil` we fall back onto the `APIVersion` domain. 

But, if the `APIVersion` domain is `nil`, it is also unclear which strategy we should follow. Since we expect the domain to be fetched the same way as the API version (mentioned earlier). Should we fallback onto the legacy endpoints? Or just return a `nil` request? 

If we fallback, should it only be when federation is disabled? Or in both cases?

### Testing

Updated tests to:
- Check that the path is set correctly for different API versions
- Set api version when needed

----
#### PR Post Submission Checklist for internal contributors (Optional)

----
#### PR Post Merge Checklist for internal contributors

----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.
